components/places/src/sync/history/store.rs

```diff
@@ -0,0 +1,184 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use std::result;
+use std::cell::Cell;
+use std::ops::Deref;
+use error::*;
+use failure;
+use rusqlite::{Connection};
+use rusqlite::{types::{ToSql, FromSql}};
+use serde_json;
+use sql_support::{ConnExt};
+use sync15_adapter::{
+    GlobalState,
+    GlobalStateProvider,
+    ClientInfo,
+    IncomingChangeset,
+    OutgoingChangeset,
+    Store,
+    ServerTimestamp
+};
+use sync15_adapter::request::{CollectionRequest};
+
+use super::record::{HistorySyncRecord};
+use super::incoming_plan::{IncomingPlan, plan_incoming_record};
+use super::super::super::storage::history_sync::{apply_synced_visits} ;
+
+
+static LAST_SYNC_META_KEY:    &'static str = "history_last_sync_time";
+static GLOBAL_STATE_META_KEY: &'static str = "history_global_state";
+
+// Lifetime here seems wrong
+pub struct HistoryStore<'a> {
+    pub db: &'a Connection,
+    pub client_info: Cell<Option<ClientInfo>>,
+}
+
+impl<'a> HistoryStore<'a> {
+    pub fn new(db: &'a Connection) -> Self {
+        Self { db, client_info: Cell::new(None) }
+    }
+
+    fn put_meta(&self, key: &str, value: &ToSql) -> Result<()> {
+        self.execute_named_cached(
+            "REPLACE INTO moz_meta (key, value) VALUES (:key, :value)",
+            &[(":key", &key as &ToSql), (":value", value)]
+        )?;
+        Ok(())
+    }
+
+    fn get_meta<T: FromSql>(&self, key: &str) -> Result<Option<T>> {
+        Ok(self.try_query_row(
+            "SELECT value FROM moz_meta WHERE key = :key",
+            &[(":key", &key as &ToSql)],
+            |row| Ok::<_, Error>(row.get_checked(0)?),
+            true
+        )?)
+    }
+
+    fn do_apply_incoming(
+        &self,
+        inbound: IncomingChangeset
+    ) -> Result<OutgoingChangeset> {
+        // for a first-cut, let's do this in the most naive way possible...
```

Comment:
I feel like it's easy enough to use a transaction here that there isn't really a reason not to.